provider/pdns.go

```diff
@@ -0,0 +1,369 @@
+		return nil, errors.New("PDNS Provider does not support domain filter")
+	}
+	// We do not support dry running, exit safely instead of surprising the user
+	// TODO: Add Dry Run support
+	if dryRun {
+		return nil, errors.New("PDNS Provider does not currently support dry-run")
+	}
+
+	if server == "localhost" {
+		log.Warnf("PDNS Server is set to localhost, this may not be what you want. Specify using --pdns-server=")
+	}
+
+	cfg := pgo.NewConfiguration()
+	cfg.Host = server
+	cfg.BasePath = server + apiBase
+
+	provider := &PDNSProvider{
+		client: &PDNSAPIClient{
+			dryRun:  dryRun,
+			authCtx: context.WithValue(context.TODO(), pgo.ContextAPIKey, pgo.APIKey{Key: apikey}),
+			client:  pgo.NewAPIClient(cfg),
+		},
+	}
+
+	return provider, nil
+}
+
+func (p *PDNSProvider) convertRRSetToEndpoints(rr pgo.RrSet) (endpoints []*endpoint.Endpoint, _ error) {
+	endpoints = []*endpoint.Endpoint{}
+
+	for _, record := range rr.Records {
+		// If a record is "Disabled", it's not supposed to be "visible"
+		if !record.Disabled {
+			endpoints = append(endpoints, endpoint.NewEndpointWithTTL(rr.Name, record.Content, rr.Type_, endpoint.TTL(rr.Ttl)))
+		}
+	}
+
+	return endpoints, nil
+}
+
+// ConvertEndpointsToZones marshals endpoints into pdns compatible Zone structs
+func (p *PDNSProvider) ConvertEndpointsToZones(eps []*endpoint.Endpoint, changetype pdnsChangeType) (zonelist []pgo.Zone, _ error) {
+	/* eg of mastermap
+	    { "example.com":
+		{ "app.example.com":
+		    { "A": ["192.168.0.1", "8.8.8.8"] }
+		    { "TXT": ["\"heritage=external-dns,external-dns/owner=example\""] }
+		}
+	    }
+	*/
+
+	zonelist = []pgo.Zone{}
+	endpoints := make([]*endpoint.Endpoint, len(eps))
+	copy(endpoints, eps)
+
+	// Sort the endpoints array so we have deterministic inserts
+	sort.SliceStable(endpoints,
+		func(i, j int) bool {
+			// We only care about sorting endpoints with the same dnsname
+			if endpoints[i].DNSName == endpoints[j].DNSName {
+				return endpoints[i].RecordType < endpoints[j].RecordType
+			}
+			return endpoints[i].DNSName < endpoints[j].DNSName
+		})
+
+	zones, _, err := p.client.ListZones()
+	if err != nil {
+		return nil, err
+	}
+
+	// Sort the zone by length of the name in descending order, we use this
+	// property later to ensure we add a record to the longest matching zone
+
+	sort.SliceStable(zones, func(i, j int) bool { return len(zones[i].Name) > len(zones[j].Name) })
+
+	// NOTE: Complexity of this loop is O(Zones*Endpoints).
+	// A possibly faster implementation would be a search of the reversed
+	// DNSName in a trie of Zone names, which should be O(Endpoints), but at this point it's not
+	// necessary.
+	for _, zone := range zones {
+		zone.Rrsets = []pgo.RrSet{}
+		for i := 0; i < len(endpoints); {
+			ep := endpoints[0]
+			dnsname := ensureTrailingDot(ep.DNSName)
+			if strings.HasSuffix(dnsname, zone.Name) {
+				// The assumption here is that there will only ever be one target
```

Comment:
The PR is against master, so that shouldn't be a problem, am I right? If so, it is not a concern.

Reply:
Indeed, this is just here as a note in case it changes in the future and results in a breakage. It's an assumption, so I thought to document it.